tests/functional/transaction-pool-api/source/snapshot.ts

```diff
@@ -0,0 +1,147 @@
+import { Contracts, Identifiers } from "@mainsail/contracts";
+import { assert, Sandbox } from "@mainsail/test-framework";
+import { BigNumber } from "@mainsail/utils";
+
+import { getWalletByAddressOrPublicKey } from "./utils.js";
+
+export const takeSnapshot = async (sandbox: Sandbox): Promise<Snapshot> => {
+	const snapshot = new Snapshot(sandbox);
+	const { walletRepository } = sandbox.app.get<Contracts.State.Service>(Identifiers.State.Service).getStore();
+
+	for (const wallet of walletRepository.allByPublicKey()) {
+		await snapshot.add(wallet.getPublicKey()!);
+	}
+
+	for (const wallet of walletRepository.allByAddress()) {
+		await snapshot.add(wallet.getAddress()!);
+	}
+
+	return snapshot;
+};
+
+export class Snapshot {
+	private balances: Record<string, BigNumber> = {};
+
+	public constructor(public sandbox: Sandbox) {}
+
+	public async add(addressOrPublicKey: string): Promise<void> {
+		const wallet = await getWalletByAddressOrPublicKey(this.sandbox, addressOrPublicKey);
+		this.balances[wallet.getAddress()] = wallet.getBalance();
+	}
+
+	public async validate(): Promise<void> {
+		const { walletRepository } = this.sandbox.app
+			.get<Contracts.State.Service>(Identifiers.State.Service)
+			.getStore();
+
+		// All balance changes from block 1 onwards
+		const balanceDeltas = await this.collectBalanceDeltas();
+
+		// Verify final balance of all wallets matches with delta and snapshot taken at block 0
+		const seenAddresses = new Map<string, boolean>();
```

Comment:
seenAddresses can be removed, because it is just caching a data. If you use only allByAddress  loop then wallet is checked exactly once.